java/ql/src/Security/CWE/CWE-90/LdapInjectionLib.qll

```diff
@@ -0,0 +1,441 @@
+
+/** A data flow sink for unvalidated user input that is used to construct LDAP queries. */
+abstract class LdapInjectionSink extends DataFlow::ExprNode { }
+
+/**
+ * A taint-tracking configuration for unvalidated user input that is used to construct LDAP queries.
+ */
+class LdapInjectionFlowConfig extends TaintTracking::Configuration {
+  LdapInjectionFlowConfig() { this = "LdapInjectionFlowConfig" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof LdapInjectionSource }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof LdapInjectionSink }
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    node.getType() instanceof PrimitiveType or node.getType() instanceof BoxedType
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) {
+    ldapNameStep(node1, node2) or
+    ldapNameAddAllStep(node1, node2) or
+    ldapNameGetCloneStep(node1, node2) or
+    filterStep(node1, node2) or
+    filterToStringStep(node1, node2) or
+    unboundIdSearchRequestStep(node1, node2) or
+    unboundIdSearchRequestDuplicateStep(node1, node2) or
+    unboundIdSearchRequestSetStep(node1, node2) or
+    ldapQueryStep(node1, node2) or
+    ldapQueryBaseStep(node1, node2) or
+    ldapQueryBuilderStep(node1, node2) or
+    hardcodedFilterStep(node1, node2) or
+    springLdapFilterToStringStep(node1, node2) or
+    ldapNameBuilderStep(node1, node2) or
+    ldapNameBuilderBuildStep(node1, node2) or
+    ldapUtilsStep(node1, node2) or
+    apacheSearchRequestStep(node1, node2) or
+    apacheSearchRequestGetStep(node1, node2) or
+    apacheLdapDnStep(node1, node2) or
+    apacheLdapDnGetStep(node1, node2)
+  }
+}
+
+/** A source of remote user input. */
+class RemoteSource extends LdapInjectionSource {
+  RemoteSource() { this instanceof RemoteFlowSource }
+}
+
+/** A source of local user input. */
+class LocalSource extends LdapInjectionSource {
+  LocalSource() { this instanceof LocalUserInput }
+}
+
+/**
+ * JNDI sink for LDAP injection vulnerabilities, i.e. 1st (DN) or 2nd (filter) argument to
+ * `search` method from `DirContext`.
+ */
+class JndiLdapInjectionSink extends LdapInjectionSink {
+  JndiLdapInjectionSink() {
+    exists(MethodAccess ma, Method m, int index |
+      ma.getMethod() = m and
+      ma.getArgument(index) = this.getExpr()
+    |
+      m.getDeclaringType().getAnAncestor() instanceof TypeDirContext and
+      m.hasName("search") and
+      index in [0 .. 1]
+    )
+  }
+}
+
+/**
+ * UnboundID sink for LDAP injection vulnerabilities,
+ * i.e. LDAPConnection.search, LDAPConnection.asyncSearch or LDAPConnection.searchForEntry method.
+ */
+class UnboundIdLdapInjectionSink extends LdapInjectionSink {
+  UnboundIdLdapInjectionSink() {
+    exists(MethodAccess ma, Method m, int index, Parameter param |
+      ma.getMethod() = m and
+      ma.getArgument(index) = this.getExpr() and
+      m.getParameter(index) = param
```

Comment:
These lines are essentially repeated in each of the `LdapInjectionSink` subclasses.
I'd suggest the following refactor: Seeing as all of the various `LdapInjectionSink`s follow a common pattern of being an argument in a specific position to a specific method, I think it would be nicer to collect all those method-index pairs in a predicate like this:
```
predicate ldapInjectionSinkMethod(Method m, int index) {
  m.getDeclaringType().getAnAncestor() instanceof TypeDirContext and
  m.hasName("search") and
  index in [0 .. 1]
  or
  exists(Parameter param | m.getParameter(index) = param and not param.isVarargs() |
    m instanceof MethodUnboundIdLDAPConnectionSearch or
    m instanceof MethodUnboundIdLDAPConnectionAsyncSearch or
    m instanceof MethodUnboundIdLDAPConnectionSearchForEntry
  )
  or
  // etc...
}
```
Then the actual `LdapInjectionSink` can simply be defined as
```
class LdapInjectionSink extends DataFlow::ExprNode {
  LdapInjectionSink() {
    exists(MethodAccess ma, Method m, int index |
      ma.getMethod() = m and
      ma.getArgument(index) = this.getExpr() and
      ldapInjectionSinkMethod(m, index)
    )
  }
}
```

Reply:
Note also here that I've made use of the `isVarargs()` predicate directly on the parameter, which should also be a simplification and remove the need for `predicate isVarargs(Callable c, int paramIndex)`.

Reply:
I refactored as you suggested.
I created separate `*ldapInjectionSinkMethod` methods for JNDI, UnboundID, Spring and Apache LDAP.
I also refactored `isVarargs` usage and removed the predicate.